File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.py
import sys

input = sys.stdin.readline

#내려갈때마다 대각선 왼쪽 or 대각선 오른쪽만 선택
#현재층의 i번째 인덱스면 다음층의 i, i+1 둘중하나 선택가능
#현재의 선택이 미래에 영향. 그리디 x dp로 풀어야함

n = int(input())
dp = [[0] * 501 for _ in range(n+1)]
data = [[0]]
for _ in range(n):
    temp = list(map(int, input().split()))
    temp.insert(0,0)
    data.append(temp)

dp[1][1] = data[1][1] #1층 1번 인덱스 자명한건 미리 초기화
res = 0
for i in range(2,n+1): #2번째 층부터 탐색.
    for j in range(1,i+1):
        if j == 1:
            dp[i][j] = dp[i-1][1] + data[i][1]
        elif j == i:
            dp[i][j] = dp[i-1][j-1] + data[i][j]
        else:
            caseA = dp[i-1][j-1] + data[i][j]
            caseB = dp[i-1][j] + data[i][j]
            dp[i][j] = max(caseA, caseB)

for i in range(1,n+1):
    res = max(dp[n][i], res)
    
print(res)

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static int n;
    static int[][] data, dp;
    // ! 아래층에 있는 수는 현재 층에서 선택된 수의 대각선 왼쪽 또는 대각선 오른쪽에 있는 것 중에서만 선택 가능
    public static void main(String[] args) throws IOException {
        n = Integer.parseInt(br.readLine());
        // 현재의 선택이 나중의 선택에 영향을 끼침 -> dp
        data = new int[n][];
        dp = new int[n][];
        for (int i = 0; i < n; i++) {
            data[i] = new int[i + 1];
            dp[i] = new int[i + 1];
        }

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j <= i; j++) {
                int val = Integer.parseInt(st.nextToken());
                data[i][j] = val;
            }
        }

        dp[0][0] = data[0][0];  // 자명한 값
        for (int i = 1; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                if(j == 0) dp[i][j] = dp[i - 1][j] + data[i][j];
                else if(j == i) dp[i][j] = dp[i - 1][j - 1] + data[i][j];
                else dp[i][j] = Math.max(dp[i-1][j-1], dp[i-1][j]) + data[i][j];
            }
        }

        int res = 0;
        for (int i = 0; i < n; i++) {
            res = Math.max(res, dp[n - 1][i]);
        }
        System.out.println(res);
    }
}